// node-cli/lib/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.comment = exports.success = exports.step = exports.question = exports.info = exports.warn = exports.error = void 0;
const elements_1 = require("./elements");
const fragment_1 = require("./jsx/fragment");
const core_1 = require("./jsx/core");
const symbols_1 = require("./symbols");
const cli_1 = require("./cli");
const arrow = core_1.h(elements_1.Blue, null, symbols_1.symbols.arrow);
const questionMark = core_1.h(elements_1.Green, null, symbols_1.symbols.question);
const yesNo = core_1.h(elements_1.Gray, null, symbols_1.symbols.yesNo);
const squareRoot = core_1.h(elements_1.Green, null, symbols_1.symbols.squareRoot);
const warning = core_1.h(elements_1.Yellow, null, symbols_1.symbols.warning);
const information = core_1.h(elements_1.Cyan, null, symbols_1.symbols.info);
const error = (msg, exit) => {
    const symbol = core_1.h(elements_1.Red, null, symbols_1.symbols.cross);
    if (msg)
        cli_1.write(`${symbol} ${msg}`);
    else
        cli_1.write(`${symbol} unknown error`);
    if (exit)
        process.exit(1);
};
exports.error = error;
const warn = (text) => {
    cli_1.write(core_1.h(elements_1.Bold, null,
        warning,
        " ",
        text));
};
exports.warn = warn;
const info = (text) => {
    cli_1.write(core_1.h(elements_1.Bold, null,
        information,
        " ",
        text));
};
exports.info = info;
const question = (text, comment, output = true) => {
    let message = '';
    message += (core_1.h(fragment_1.Fragment, null,
        core_1.h(elements_1.Bold, null, `${questionMark} ${text} `),
        comment ? core_1.h(elements_1.Gray, null, comment) : null));
    if (output)
        cli_1.write(message);
    return message;
};
exports.question = question;
const step = (msg, output = true) => {
    const message = `${arrow} ${msg}`;
    if (output)
        cli_1.write(message);
    return message;
};
exports.step = step;
const success = (text, comment = null, output = true) => {
    const msg = (core_1.h(elements_1.Bold, null,
        squareRoot,
        " ",
        text));
    const cmt = core_1.h(elements_1.Gray, null, comment);
    let message = '';
    if (comment)
        message = `${msg} ${cmt}`;
    else
        message = msg;
    if (output)
        cli_1.write(message);
    return message;
};
exports.success = success;
const comment = (text, output = true) => {
    const message = core_1.h(elements_1.Gray, null, text);
    if (output)
        cli_1.write(message);
    return message;
};
exports.comment = comment;
